amplify/backend/api/colonycdapp/schema.graphql

```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
   toDomain: Domain @hasOne(fields: ["toDomainId"])
   fundamentalChainId: Int
   newColonyVersion: Int
+  roles: ColonyActionRoles
+
+  # Required since some actions might have multiple event entries in the list
+  # which need to be iterated over and exposed individually
+  #
+  # This should be a list/array, but this is yet another thing Amplify is bad at
+  # The official response from them is to use DynamoDB directly:
+  # https://stackoverflow.com/questions/50992106/append-item-to-list-using-aws-appsync-to-dynamodb
+  # Which is just freaking great!
+  #
+  # As such, I've decided to simplify this, an just pass a stringified JSON around
+  # since this is for display purpouses only.
+  #
+  # If anyone has a better idea on how to handle this, I'll gladly change it!
+  #
+  # When / if you decide to use it, just be mindful of what format other people
+  # are using, and follow suit.
+  #
+  # Roles Types:
+  # [{ id: String, type: String, role: Number, setTo: boolean }]
+  individualEvents: String
+}
+
+type ColonyActionRoles {
+  role_0: Boolean # recovery
+  role_1: Boolean # root
+  role_2: Boolean # arbitration
+  role_3: Boolean # architecture
+  role_5: Boolean # funding
+  role_6: Boolean # administration
+}
+
+# Roles have been set as role_<number> so we can more easily map them to their
+# actual contract counterpart on the client and block ingestor side(s)
+
+type ColonyRole @model {
+  id: ID! @index # colonyAddress_domainNativeId_userAddress_roles
+  domain: Domain! @hasOne
+
+  targetAddress: ID # Amplify will automatically populate one of the following fields with related model if it finds one
+  targetUser: User @hasOne(fields: ["targetAddress"])
+  targetColony: Colony @hasOne(fields: ["targetAddress"])
+  targetExtension: ColonyExtension @hasOne(fields: ["targetAddress"])
+  targetToken: Token @hasOne(fields: ["targetAddress"])
+
+  latestBlock: Int!
+  role_0: Boolean # recovery
+  role_1: Boolean # root
+  role_2: Boolean # arbitration
+  role_3: Boolean # architecture
+  role_5: Boolean # funding
+  role_6: Boolean # administration
+}
+
+# it's a model, not a type since this is designed to work outside of the main
+# colony model, or colony role model, as, by design it should be only used
+# in very specific cases
+#
+# This should take a snapshot of the user's full roles/permissions at a specific block
+
+type ColonyHistoricRole @model {
+  id: ID! # colonyAddress_domainNativeId_userAddress_blockNumber_roles
+
+  # I HATE AMPLIFY SORTING! MAY IT DIE A THOUSAND DEATHS!
+  #
+  # @NOTE Always set to `type` to "SortedHistoricRole" to enable Amplify sorting, it DOES NOT WORK with `id`
+  #
+  # Apparently, it can only filter non-primary fields, meaning that if we sort by `blockNumber` then that
+  # field is considered primary, so it won't allow us to filter by it.
```

Comment:
What is the intended access pattern here? 

If you wanted to get permissions at an exact block number, you could also have block number as an index.

Or if you wanted to get all permissions for a given target address, index the address and then you can have the block number as the sort key. 

We _probably_ don't need to add an abritrary field just to enable the desired access pattern (although hard to comment for sure since I'm not sure what it is).

Reply:
If you indexed the targetAddress, and added the block number as a sortKeyField, then you'd be able to get all roles at the targetAddress and the block number without having to do any filtering. 

Then you'd only have to filter by colonyAddress to enable this pattern:

> get the permissions for user X at block Y within colony Z

This would also allow to you to get Roles for a target address from a  range block numbers since the sort key enables a variety of comparisons (gt, lt, gt & lt etc)